src/js/_enqueues/lib/emoji-loader.js

```diff
@@ -139,82 +246,169 @@
 		var script = document.createElement( 'script' );
-
 		script.src = src;
-		script.defer = script.type = 'text/javascript';
-		document.getElementsByTagName( 'head' )[0].appendChild( script );
+		script.defer = true;
+		document.head.appendChild( script );
 	}
 
-	tests = Array( 'flag', 'emoji' );
-
 	settings.supports = {
 		everything: true,
 		everythingExceptFlag: true
 	};
 
-	/*
-	 * Tests the browser support for flag emojis and other emojis, and adjusts the
-	 * support settings accordingly.
-	 */
-	for( ii = 0; ii < tests.length; ii++ ) {
-		settings.supports[ tests[ ii ] ] = browserSupportsEmoji( tests[ ii ] );
-
-		settings.supports.everything = settings.supports.everything && settings.supports[ tests[ ii ] ];
+	// Obtain the emoji support from the browser, asynchronously when possible.
+	var sessionSupportsPromise = new Promise( function ( resolve ) {
+		var sessionSupports = getSessionSupports();
+		if ( sessionSupports ) {
+			resolve( sessionSupports );
+			return;
+		}
 
-		if ( 'flag' !== tests[ ii ] ) {
-			settings.supports.everythingExceptFlag = settings.supports.everythingExceptFlag && settings.supports[ tests[ ii ] ];
+		if ( supportsWorkerOffloading() ) {
+			try {
+				/*
+				 * Note that this string contains the real source code for the
+				 * copied functions, _not_ a string representation of them. This
+				 * is because it's not possible to transfer a Function across
+				 * threads. The lack of quotes is intentional. The function names
+				 * are copied to variable names since minification will munge the
+				 * function names, thus breaking the ability for the functions to
+				 * refer to each other.
+				 */
+				var workerScript =
+					'var emojiSetsRenderIdentically = ' + emojiSetsRenderIdentically + ';' +
+					'var browserSupportsEmoji = ' + browserSupportsEmoji + ';' +
+					'var testEmojiSupports = ' + testEmojiSupports + ';' +
+					'postMessage(testEmojiSupports(' + JSON.stringify(tests) + '));';
+				var blob = new Blob([workerScript], {
+					type: 'text/javascript'
+				});
+				var worker = new Worker(URL.createObjectURL(blob));
+				worker.onmessage = function (event) {
+					sessionSupports = event.data;
+					setSessionSupports(sessionSupports);
+					resolve(sessionSupports);
+				};
+				return;
+			} catch ( e ) {}
 		}
-	}
 
-	settings.supports.everythingExceptFlag = settings.supports.everythingExceptFlag && ! settings.supports.flag;
+		sessionSupports = testEmojiSupports( tests );
+		setSessionSupports( sessionSupports );
+		resolve( sessionSupports );
+	} );
 
-	// Sets DOMReady to false and assigns a ready function to settings.
-	settings.DOMReady = false;
-	settings.readyCallback = function() {
-		settings.DOMReady = true;
-	};
+	// Create a promise for DOMContentLoaded since the worker logic may finish after the event has fired.
+	var domReadyPromise = new Promise( function ( resolve ) {
+		document.addEventListener( 'DOMContentLoaded', resolve, {
+			once: true
+		} );
+	} );
```

Comment:
this is also not a big deal, but it's curious that we interrupt the `sessionSupportsPromise` to create the `domReadyPromise` here. if we put `domReadyPromise` above `sessionSupportsPromise` couldn't we eliminate the creation of a variable for `sessionSupportsPromise` entirely and leave it as a single continuous chain instead of trapping a promise and then "re-starting" it in the code later?

```js
var domReadyPromise = new Promise( … );

new Promise( function ( resolve ) {
	…
} )
	.then( function ( sessionSupports ) {

	} )
	.then( function () {
		return domReadyPromise;
	} )
	.then( function () {
		…
	} );
```

Reply:
> this is also not a big deal, but it's curious that we interrupt the `sessionSupportsPromise` to create the `domReadyPromise` here.

I thought maybe it was easier to read, but apparently not. So I've made the change in aaf5718db58582e5b1bc8b9fca1be49adf20224e

Reply:
> There are try/catch blocks around code that I expect might throw errors.

makes sense. I guess we have to trust that the code inside those Promises won't throw. if for some reason they would (such as the unexpected bug with invalid code points) we would get an error about an unhandled catch, I think.

> I don't know if there should be a different outcome?

maybe the only outcome is swallowing the errors and doing any necessary cleanup. if leaving session storage untouched is all the cleanup that's required, that makes sense. I have assumed that the failure mode is to make sure that `twemoji` and other scripts load, if we cannot determine that they aren't needed.

Reply:
The current emoji loader logic in trunk doesn't have any such error handling to force Twemoji to load in the case of an exception while gathering settings. I think we can leave that for future consideration.